python_modules/dagster/dagster/_core/definitions/op_invocation.py

```diff
@@ -59,31 +121,24 @@ def op_invocation_result(
     context = context or build_op_context()
 
     resource_arg_mapping = {arg.name: arg.name for arg in compute_fn.get_resource_args()}
-    resource_args_from_kwargs = {}
-    for resource_arg in resource_arg_mapping:
-        if resource_arg in kwargs:
-            resource_args_from_kwargs[resource_arg] = kwargs[resource_arg]
-            del kwargs[resource_arg]
 
-    resources_provided_in_multiple_places = (resource_args_from_kwargs) and (context.resource_keys)
+    # Extract any resource inputs from the args and kwargs
+    extracted = _extract_resource_args(compute_fn, args, kwargs, resource_arg_mapping)
+
+    adjusted_args = extracted.args
+    adjusted_kwargs = extracted.kwargs
+    resource_inputs_from_args = extracted.resource_inputs
+
+    resources_provided_in_multiple_places = (resource_inputs_from_args) and (context.resource_keys)
     if resources_provided_in_multiple_places:
         raise DagsterInvalidInvocationError("Cannot provide resources in both context and kwargs")
 
-    if resource_args_from_kwargs:
-        context = context.replace_resources(resource_args_from_kwargs)
+    if resource_inputs_from_args:
+        context = context.replace_resources(resource_inputs_from_args)
 
-    try:
-        bound_context = context.bind(op_def_or_invocation)
-    except DagsterInvalidDefinitionError as e:
-        if any(isinstance(arg, ResourceDefinition) for arg in args):
-            raise DagsterInvalidInvocationError(
-                str(e)
-                + "\n\nIf directly invoking an op/asset, you may not provide resources as"
-                " positional"
-                " arguments, only as keyword arguments."
-            ) from e
-        raise
-    input_dict = _resolve_inputs(op_def, args, kwargs, bound_context)
+    bound_context = context.bind(op_def_or_invocation)
+
+    input_dict = _resolve_inputs(op_def, adjusted_args, adjusted_kwargs, bound_context)
 
     result = invoke_compute_fn(
```

Comment:
I think this codepath would be a lot clearer if we renamed the `kwargs` argument in `invoke_compute_fn` to "input_by_param_name" or something. Because what is actually happening here is that post this workstream `invoke_compute_fn` is constructing the `kwargs` rather than getting them passed to.